Docs Improvements (#169)
awslabs/amazon-s3-find-and-forget: docs/ARCHITECTURE.md

in S3 buckets. In order to achieve these goals the solution has adopted the
following design principles:

> 1. **Secure by design:** Every component is implemented with least privilege

Reviewer: I still feel this would be better presented as a list in the form: 

Every component is implemented according to the following standard:
- x
- y
- z

Author: Done

> 1. **Security by design:** Every component implements least privilege access,

Reviewer 2: ```suggestion
1. **Secure by design:** Every component is implemented with least privilege access,
```

1. Noun ➡️ adjective form.
2. Remove "every component" from being the implementer.

   encryption is performed at all layers at rest and in transit, authentication
   is provided out of the box, expiration of logs is configurable, and record
   identifiers (known as **Match IDs**) are automatically obfuscated or
   irreversibly deleted as soon as possible when persisting state.

> 2. **Built to scale:** The system has been designed and battle tested for

Reviewer 2: ```suggestion
2. **Built to scale:** The system has been designed and tested for
```

Opinionated suggestion here, but "battle" is unnecessary :)

> 2. **Built to scale:** The system has been designed and tested for performance

Reviewer: ```suggestion
2. **Built to scale:** The system is designed and tested for performance
```

Reviewer: also _designed and tested for performance and high availability with Data Lakes of the..._ - rereading this now I don't know if that actually makes sense. The scale of the data lake doesn't really impact availability? And the statement doesn't specify the standard of performance. I think we should be saying that _the system is designed and tested to support up to petabyte scale data lakes_ or something similar?

>    performance and high availability with Data Lakes of TB/PB size.

Reviewer 3: ```suggestion
   performance and high availability with Data Lakes of petabyte size.
```

   and high availability with Data Lakes of the scale of petabyte size.
3. **Cost optimised:**
   - **Perform work in batches:** Since the time complexity of removing a single
     vs multiple records in a single object is practically equal and it is
     common for data owners to have the requirement of removing data within a
     given _timeframe_, the solution is designed to allow the solution operator
     to "queue" multiple matches to be removed in a single job.
   - **Fail fast:** A deletion job takes place in two distinct phases: Find and
     Forget. The Find phase queries the objects in your S3 data lakes to find
     any objects which contain records where a specified column contains at
     least one of the Match IDs in the deletion queue. If any queries fail, the
     job will abandon as soon as possible and the Forget phase will not take
     place. The Forget Phase takes the list of objects returned from the Find
     phase, and deletes only the relevant rows in those objects.
   - **Optimised for Parquet:** The process of separating the two phases, as

>      opposite of processing the read/write on an object basis, is another cost

Reviewer: ```suggestion
     opposed to processing the read/write on an object basis, is another cost
```

     allow to efficiently find matches by reading only relevant columns instead
     of full content from all columns. By processing only the relevant columns
     during the Find and delegating the queries to a serverless managed service
     such as Amazon Athena, we can then process full rows only during the Forget
     phase, restricting the most expensive operation to only the relevant

>      objects.

Reviewer 2: I find this point is a bit difficult to follow - there's a lot of info including some bits that are perhaps not necessary (e.g. describing the approach _not taken_). I don't have an immediate suggested edit, but I think simplifying the sentence to only make a few points would make it more readable.

Reviewer 2: ```
    - **Optimised for Parquet:** The split phase approach optimises scanning
      for columnar dense formats such as Parquet.  The Find phase only
      retrieves and processes the data for relevant columns when determining
      which S3 objects need to be processed in the Forget phase. This
      approach can have significant cost savings when operating on large data
      lakes with sparse matches.
```

How about this?

   - **Serverless:** Where possible, the solution only uses Serverless

>      components, in order to avoid paying for idle resources. All the components

Reviewer 2: ```suggestion
     components to avoid paying for idle resources. All the components
```

     for Web UI, API and Deletion Jobs are Serverless (for more information
     consult the [Cost Overview guide].
4. **Robust monitoring and logging:** When performing deletion jobs, information

>    is provided in real-time to allow good observability. After the job

Reviewer 2: ```suggestion
   is provided in real-time to provide visibility. After the job
```

   completes, detailed reports are available documenting all the actions
   performed to individual S3 Objects, and detailed error traces in case of

>    failures to guide the user to the troubleshooting process and identify

Reviewer 2: ```suggestion
   failures to facilitate troubleshooting processes and identify
```
